server/src/test/java/org/elasticsearch/search/vectors/VectorDataTests.java

```diff
@@ -0,0 +1,172 @@
+    }
+
+    public void testShouldThrowForOutsideRangeWhenConvertingToByte() {
+        float[] vec = new float[] { 1f, 2f, 200f };
+
+        VectorData vectorData = new VectorData(vec, null);
+        expectThrows(IllegalArgumentException.class, vectorData::asByteVector);
+    }
+
+    public void testEqualsAndHashCode() {
+        VectorData v1 = new VectorData(new float[] { 1, 2, 3 }, null);
+        VectorData v2 = new VectorData(null, new byte[] { 1, 2, 3 });
+        assertNotEquals(v1, v2);
+        assertNotEquals(v1.hashCode(), v2.hashCode());
+
+        VectorData v3 = new VectorData(null, new byte[] { 1, 2, 3 });
+        assertEquals(v2, v3);
+        assertEquals(v2.hashCode(), v3.hashCode());
+    }
+
+    public void testParseHexCorrectly() throws IOException {
+        VectorData expected = new VectorData(null, new byte[] { 64, 10, -30, 10 });
+        XContentParser parserMock = mock(XContentParser.class);
+        when(parserMock.currentToken()).thenReturn(XContentParser.Token.VALUE_STRING);
+        when(parserMock.text()).thenReturn("400ae20a");
+        VectorData parsed = VectorData.parseXContent(parserMock);
+        assertEquals(expected, parsed);
+        UnsupportedOperationException ex = expectThrows(UnsupportedOperationException.class, parsed::asFloatVector);
+        assertThat(ex.getMessage(), containsString("cannot convert to float, as we're explicitly using a byte vector"));
+    }
+
+    public void testParseFloatArray() throws IOException {
+        VectorData expected = new VectorData(new float[] { 1f, -1f, .1f }, null);
+        XContentParser parserMock = mock(XContentParser.class);
+        when(parserMock.currentToken()).thenReturn(XContentParser.Token.START_ARRAY);
+        when(parserMock.nextToken()).thenReturn(
+            XContentParser.Token.VALUE_NUMBER,
+            XContentParser.Token.VALUE_NUMBER,
+            XContentParser.Token.VALUE_NUMBER,
+            XContentParser.Token.END_ARRAY
+        );
+        when(parserMock.floatValue()).thenReturn(1f, -1f, .1f);
+        VectorData parsed = VectorData.parseXContent(parserMock);
+        assertEquals(expected, parsed);
+    }
+
+    public void testParseByteArray() throws IOException {
+        byte[] expectedByteArray = new byte[] { 64, 10, -30, 10 };
+        XContentParser parserMock = mock(XContentParser.class);
+        when(parserMock.currentToken()).thenReturn(XContentParser.Token.START_ARRAY);
+        when(parserMock.nextToken()).thenReturn(
+            XContentParser.Token.VALUE_NUMBER,
+            XContentParser.Token.VALUE_NUMBER,
+            XContentParser.Token.VALUE_NUMBER,
+            XContentParser.Token.VALUE_NUMBER,
+            XContentParser.Token.END_ARRAY
+        );
+        when(parserMock.floatValue()).thenReturn(64f, 10f, -30f, 10f);
+        VectorData parsed = VectorData.parseXContent(parserMock);
+        assertArrayEquals(expectedByteArray, parsed.asByteVector());
+    }
```

Comment:
```suggestion
    public void testParseByteArray() throws IOException {
        byte[] expectedByteArray = new byte[] { 64, 10, -30, 10 };
        String toParse = "[64,10,-30,10]";
        try (XContentParser parser = XContentHelper.createParserNotCompressed(
            XContentParserConfiguration.EMPTY,
            new BytesArray(toParse),
            XContentType.JSON
        )){
            parser.nextToken();
            VectorData parsed = VectorData.parseXContent(parser);
            assertArrayEquals(expectedByteArray, parsed.asByteVector());
        }
    }
```
Lets try to not use mocks if possible. Here is an example of how to transform this particular test to not use mocks.

Reply:
Addressed in 2c346822d6d66feab114538f4dd9e2b1313251f0